### ISSUE TITLE
Always receive usage printout from java program when running YCSB over JDBC

### ISSUE THREAD

**Issue by petergaultney** [brianfrankcooper/YCSB#407]
```
bin/ycsb load jdbc -P workloads/workloada -cp /usr/share/java/*.jar -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root
```

always results in 
```
Usage: java com.yahoo.ycsb.Client [options]
Options:
  -threads n: execute using n threads (default: 1) - can also be specified as the 
        "threadcount" property using -p
  -target n: attempt to do n operations per second (default: unlimited) - can also
       be specified as the "target" property using -p
  -load:  run the loading phase of the workload
  -t:  run the transactions phase of the workload (default)
  -db dbname: specify the name of the DB to use (default: com.yahoo.ycsb.BasicDB) - 
        can also be specified as the "db" property using -p
  -P propertyfile: load properties from the given file. Multiple files can
           be specified, and will be processed in the order specified
  -p name=value:  specify a property to be passed to the DB and workloads;
          multiple properties can be specified, and override any
          values in the propertyfile
  -s:  show status during run (default: no status)
  -l label:  use label for status (e.g. to label one experiment out of a whole batch)

Required properties:
  workload: the name of the workload class to use (e.g. com.yahoo.ycsb.workloads.CoreWorkload)

To run the transaction phase from multiple servers, start a separate client on each.
To run the load phase from multiple servers, start a separate client on each; additionally,
use the "insertcount" and "insertstart" properties to divide up the records to be inserted
```

I've been going around in circles on this for about 4 hours. The java program starts up and manages to load the JDBC connector, but that's as far as it goes.

I've been poking around in the source, and I can see that there are a number of places where usageMessage() may get called and print the message seen above. Unfortunately all of them also result in System.exit(0), so there's no way to tell why the program is unhappy with my supplied parameters other than by editing the code to add debug statements and recompiling.

A good first step would be for someone to add descriptive error messages to Client.java and CommandLine.java, and to reformat that code so that it's readable (the indentation is atrocious). 

I hate to come across as ungrateful for free code, but the fact is that I've so far not actually gotten a bit of use out of it despite hours of investment. For the time being, I think it would help to warn people in the README that in fact the system is not particularly stable, and to provide step-by-step instructions for a selected subset of the SQL and NoSQL database bindings that are guaranteed to work from scratch on a relatively modern Linux distribution.

**Comment by busbey** [comment]
>  For the time being, I think it would help to warn people in the README that in fact the system is not
> particularly stable, and to provide step-by-step instructions for a selected subset of the SQL and
> NoSQL database bindings that are guaranteed to work from scratch on a relatively modern Linux
> distribution.

We do list these things, but it's attached to each release (for example, [our latest 0.3.1](https://github.com/brianfrankcooper/YCSB/releases/tag/0.3.1)). Unfortunately, the JDBC driver hasn't made the list of tested datastore bindings. Sorry you had to be the one to trip over things.

I think from your command line that you're trying to use mysql?

Would linking to the release notes (rather than just a download) help set expectations? Would a version of [our page for getting started on Windows](https://github.com/brianfrankcooper/YCSB/wiki/Prerequisites-for-Windows) help address things if we had one focused on a a linux distro?

> A good first step would be for someone to add descriptive error messages to Client.java and
> CommandLine.java, and to reformat that code so that it's readable (the indentation is atrocious).

We're always happy for the contribution of better error messages. We have an open issue to fix the code formatting project wide (#267), but merging any fixes for that issue are currently waiting on us merging old PRs that would conflict (e.g. #98)

**Comment by busbey** [comment]
I saved a local copy of the [mysql example creation call](https://github.com/brianfrankcooper/YCSB/blob/master/jdbc/src/main/resources/sql/create_table.mysql) and then used it to create the table. It looks like inserting worked then:

```bash
[ycsb-0.3.1]$ mysql -u root --database ycsb < create_table.mysql
[ycsb-0.3.1]$ ./bin/ycsb load jdbc -cp /usr/share/java/mysql-connector-java.jar -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram
java -cp /usr/share/java/mysql-connector-java.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/conf:/home/busbey/ycsb/ycsb-0.3.1/conf:/home/busbey/ycsb/ycsb-0.3.1/lib/jackson-mapper-asl-1.9.4.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/core-0.3.1.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/jackson-core-asl-1.9.4.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/HdrHistogram-2.1.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-kernel-2.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/jdbc-binding-0.3.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/geronimo-jms_1.1_spec-1.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/junit-3.8.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-pool-1.5.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/geronimo-jta_1.1_spec-1.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-lib-2.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-collections-3.2.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/serp-1.13.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-lang-2.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-jdbc-2.1.1.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.JdbcDBClient -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.JdbcDBClient -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram -load
Loading workload...
Starting test.
Adding shard node URL: jdbc:mysql://localhost/ycsb
Using 1 shards
[OVERALL], RunTime(ms), 318.0
[OVERALL], Throughput(ops/sec), 12.578616352201259
[CLEANUP], Operations, 1.0
[CLEANUP], AverageLatency(us), 650.0
[CLEANUP], MinLatency(us), 650.0
[CLEANUP], MaxLatency(us), 650.0
[CLEANUP], 95thPercentileLatency(ms), 0.0
[CLEANUP], 99thPercentileLatency(ms), 0.0
[INSERT], Operations, 4.0
[INSERT], AverageLatency(us), 15788.75
[INSERT], MinLatency(us), 1260.0
[INSERT], MaxLatency(us), 59231.0
[INSERT], 95thPercentileLatency(ms), 59.0
[INSERT], 99thPercentileLatency(ms), 59.0
[INSERT], Return=0, 4
[ycsb-0.3.1]$ mysql -u root --database ycsb
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 10
Server version: 5.1.73 Source distribution

Copyright (c) 2000, 2013, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select count(*) from usertable;
+----------+
| count(*) |
+----------+
|        4 |
+----------+
1 row in set (0.00 sec)
```

Does this give you enough info to move forward?

I'd love to get a README.md set up for the jdbc binding that walks through getting this working. Any chance you could put something together?

**Comment by busbey** [comment]
Apparently I forgot to hit comment after writing this after my first comment.

If I include just the mysql connector, instead of all the jars in /usr/share, I get a different and more useful message:

```bash
[ycsb-0.3.1]$ ./bin/ycsb load jdbc -cp /usr/share/java/mysql-connector-java.jar -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram
java -cp /usr/share/java/mysql-connector-java.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/conf:/home/busbey/ycsb/ycsb-0.3.1/conf:/home/busbey/ycsb/ycsb-0.3.1/lib/jackson-mapper-asl-1.9.4.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/core-0.3.1.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/jackson-core-asl-1.9.4.jar:/home/busbey/ycsb/ycsb-0.3.1/lib/HdrHistogram-2.1.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-kernel-2.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/jdbc-binding-0.3.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/geronimo-jms_1.1_spec-1.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/junit-3.8.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-pool-1.5.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/geronimo-jta_1.1_spec-1.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-lib-2.1.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-collections-3.2.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/serp-1.13.1.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/commons-lang-2.4.jar:/home/busbey/ycsb/ycsb-0.3.1/jdbc-binding/lib/openjpa-jdbc-2.1.1.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.JdbcDBClient -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.JdbcDBClient -p db.url=jdbc:mysql://localhost/ycsb -p db.user=root -P workloads/workloada -p recordcount=4 -p measurementtype=hdrhistogram -load
Loading workload...
Starting test.
Adding shard node URL: jdbc:mysql://localhost/ycsb
Using 1 shards
Error in processing insert to table: usertablecom.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Table 'ycsb.usertable' doesn't exist
[OVERALL], RunTime(ms), 545.0
[OVERALL], Throughput(ops/sec), 0.0
[CLEANUP], Operations, 1.0
[CLEANUP], AverageLatency(us), 641.0
[CLEANUP], MinLatency(us), 641.0
[CLEANUP], MaxLatency(us), 641.0
[CLEANUP], 95thPercentileLatency(ms), 0.0
[CLEANUP], 99thPercentileLatency(ms), 0.0
[INSERT], Operations, 1.0
[INSERT], AverageLatency(us), 22120.0
[INSERT], MinLatency(us), 22112.0
[INSERT], MaxLatency(us), 22127.0
[INSERT], 95thPercentileLatency(ms), 22.0
[INSERT], 99thPercentileLatency(ms), 22.0
[INSERT], Return=-1, 1
```

So we likely have both a need for a docs improvement and some kind of bug brought on by the extra jars in /usr/share, since I got the same behavior you did on a CentOS 6.4 box.

**Comment by petergaultney** [comment]
Thanks for the response. To be honest, I want to apologize for my tone in the initial report. I was frustrated by how much time I had wasted, but probably also frustrated with myself that I hadn't just dived into the code earlier to fix my own problem.

As it turns out, my issue was even simpler - the classpath wasn't getting passed to the Java program as colon-separated values, but as space-separated, because bash was doing the globbing, so the logic in Client.java was expected roughly 4 key-value parameters and had about 700 left over.

The reason I was using /usr/share/java/*.jar is because, for reasons that I never bothered to track down, apparently just adding the mysql-connector jar to the classpath wasn't getting all of the necessary dependencies included. It would've been easier for me to debug this if the bin/ycsb script was willing to use some external environment variable. However, I realize that the reason the classpath is kept clean is to keep old versions from creeping in - when I tried to run YCSB against Cassandra and supplied the same /usr/share/java/\*.jar (this time with the escape sequence to keep bash from doing the globbing), YCSB couldn't start up the SLF4J logger because an older version was present in /usr/share/java and it was overriding the newer version present in the maven dependencies. 

For the record, I'm running all of this on a 64 bit Ubuntu 15.04 VM.

In any case, I threw in my own halfway-descriptive error messages for Client (I think similar ones would probably be appropriate for CommandLine.java), but I can't guarantee they accurately represent the error condition in all cases. Still, they at least revealed to me my stupid globbing error. I may submit a pull request for these.

As far as getting JDBC set up to work with YCSB, I still can't give a confident answer about all of the steps that I took to get it running, especially since I'm still using a seemingly-hacky "-cp /usr/share/java/\*.jar" because simply copying the mysql-connector jar into a local jdbc directory or including just it in the classpath didn't seem to do the trick. From memory, I think Java was still complaining about missing the JDBC classes even when the connector JAR alone was added to the classpath.

However, I'm hoping to continue this benchmarking, so if I run across ways in which I think I can be helpful, I'll do what I can. Having just written my own generic benchmarker for messaging queues, I was grateful to see that someone else had already done this for databases (a more complex benchmark scenario in my opinion), and I'd like to see YCSB keep up with current database technology as it is evolving too quickly for anyone to make reasonable choices without having a chance to simulate their own workload.

**Comment by busbey** [comment]
> Thanks for the response. To be honest, I want to apologize for my tone in the initial report. I was
> frustrated by how much time I had wasted, but probably also frustrated with myself that I hadn't just
> dived into the code earlier to fix my own problem.

No worries, we all get frustrated from time to time. Thanks for the apology!

> As it turns out, my issue was even simpler - the classpath wasn't getting passed to the Java program
> as colon-separated values, but as space-separated, because bash was doing the globbing, so the logic
> in Client.java was expected roughly 4 key-value parameters and had about 700 left over.

Ah! Of course, once you say it it seems obvious. This sounds like a failure of our command line parsing though, surely all of those bare paths don't look like any of our legitimate flags and we could raise an error.

> In any case, I threw in my own halfway-descriptive error messages for Client (I think similar ones would 
> probably be appropriate for CommandLine.java), but I can't guarantee they accurately represent the 
> error condition in all cases. Still, they at least revealed to me my stupid globbing error. I may submit a 
> pull request for these.

that'd be great.

> As far as getting JDBC set up to work with YCSB, I still can't give a confident answer about all of the
> steps that I took to get it running, especially since I'm still using a seemingly-hacky "-cp 
> /usr/share/java/*.jar" because simply copying the mysql-connector jar into a local jdbc directory or 
> including just it in the classpath didn't seem to do the trick. From memory, I think Java was still 
> complaining about missing the JDBC classes even when the connector JAR alone was added to the 
> classpath.

On the off chance that it helps someone track things down, the version that worked for me was the mysql-connector-java 5.1.17.

**Comment by risdenk** [comment]
@busbey @petergaultney Can this issue be closed since YCSB 0.7.0 was released and includes #408

**Comment by busbey** [comment]
we have the added error handling from #408 and the examples in the readme for JDBC now specify just the connector jar in a -cp arg. I'm satisfied, presuming @petergaultney doesn't have any additional improvements he'd like to see related to this issue.

**Comment by risdenk** [comment]
Closing this issue due to inactivity and thinking its resolved. @petergaultney if your concerns weren't addressed please reopen. Thanks!
